pkg/tools/fifosync/mutex.go

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package fifosync provides FIFO order guaranteed synchronization primitives
+package fifosync
+
+import (
+	"sync"
+)
+
+// Mutex is a FIFO order guaranteed mutex
+type Mutex struct {
+	init sync.Once
+	ch   chan struct{}
+}
+
+// Lock acquires lock or blocks until it gets free
+func (m *Mutex) Lock() {
+	m.init.Do(func() {
+		m.ch = make(chan struct{}, 1)
```

Comment:
I'm pretty sure this doesn't preserve order, as there are no guarantees whatsoever about what order blocking go routines will be awakened to enter the channel in.